experiment/measurer/experiment_coverage_utils.py

```diff
@@ -0,0 +1,189 @@
+        self.name_df = None
+
+    def add_function_entry(self, benchmark, fuzzer, trial_id, function,
+                           function_hits, time):
+        # pylint: disable=too-many-arguments
+        """Adds an entry to the function_df."""
+        fuzzer_id = self.name_to_id(fuzzer)
+        benchmark_id = self.name_to_id(benchmark)
+        function_id = self.name_to_id(function)
+        self.fuzzer_names[fuzzer] = fuzzer_id
+        self.benchmark_names[benchmark] = benchmark_id
+        self.function_names[function] = function_id
+
+        function_entry = [
+            benchmark_id, fuzzer_id, trial_id, time, function_id, function_hits
+        ]
+        self.function_entries.append(function_entry)
+
+    def add_segment_entry(self, benchmark, fuzzer, trial_id, file_name, line,
+                          column, time):
+        # pylint: disable=too-many-arguments
+        """Adds an entry to the segment_df."""
+        fuzzer_id = self.name_to_id(fuzzer)
+        benchmark_id = self.name_to_id(benchmark)
+        file_id = self.name_to_id(file_name)
+        self.fuzzer_names[fuzzer] = fuzzer_id
+        self.benchmark_names[benchmark] = benchmark_id
+        self.file_names[file_name] = file_id
+
+        segment_entry = [
+            benchmark_id, fuzzer_id, trial_id, time, file_id, line, column
+        ]
+        self.segment_entries.append(segment_entry)
+
+    def done_adding_entries(self):
+        """Generates the data frames from the individual entries."""
+
+        if len(self.segment_entries) == 0:
+            coverage_utils.logger.error(
+                'Finalizing, but no entries were added.')
+            return
+
+        self.segment_df = pd.DataFrame(self.segment_entries,
+                                       columns=[
+                                           "benchmark", "fuzzer", "trial",
+                                           "time", "file", "line", "column"
+                                       ])
+        self.function_df = pd.DataFrame(self.function_entries,
+                                        columns=[
+                                            "benchmark", "fuzzer", "trial",
+                                            "time", "function", "hits"
+                                        ])
+
+        name_entries = []
+        for name in self.benchmark_names:
+            name_entries.append([self.benchmark_names[name], name, "benchmark"])
+        for name in self.fuzzer_names:
+            name_entries.append([self.fuzzer_names[name], name, "fuzzer"])
+        for name in self.function_names:
+            name_entries.append([self.function_names[name], name, "function"])
+        for name in self.file_names:
+            name_entries.append([self.file_names[name], name, "file"])
+        self.name_df = pd.DataFrame(name_entries,
+                                    columns=["id", "name", "type"])
+
+    def name_to_id(self, name):  # pylint: disable=no-self-use
```

Comment:
Is this really necessary? How much diskspace are we saving here?

Reply:
According to my readings, this step reduces the disk space consumption by about 5-6 folds. We record file and function names and they are pretty long in general (generally 30+ characters since file names are relative paths and function names in benchmarks are long as well). Hence reducing it to a 7 character unique id helps reduce diskspace consumed considerably.

Also, for segments, even-though we only record newly discovered segments from the second cycle onwards, there can be up to 25k+ entries in the first cycle itself. For functions, we record the function names, and we record in every cycle irrespective of wether the function was previously covered or not.

Reply:
> According to my readings, this step reduces the disk space consumption by about 5-6 folds. 

Right, but how big is a file without this compression vs with it?
If we are going from 80 MB in size for a full experiment to 20 MB, I don't think it is worth it.

I'm also concerned with this method you've used. It isn't sound. There can be collisions here if two things happen to have the same 8 char prefix. Why can't we simply use a numerical ID?

Reply:
Also have you tried using bz2 for compression instead of gz? It generally compresses to smaller sizes but is also slower.

Reply:
Since we call this method in each thread, we use hashing to resolve names to ids as we want to have consistent IDs for the same names in different threads and hashing is a good way to do it. 

Moreover, we call `name_to_id` and `remove_redundant_duplicates` in each thread as this helps us reduce the memory consumption (as it reduces the df size frequently). We can use numerical ID but then we would have to resolve names to IDs after we collect all the data for the experiment which would increase the memory consumption considerably. While hashing there is possibility of collision of IDs, no doubt, but the probability is quite low. So, currently I'm truncating the computed hash at 7 characters but we can increase this to say 10 or 12 characters which would drop the probability of collision further.

Also for my local experiments (not full-scale), using `name_to_id` helps reduce disk space consumed by the `*.csv.gz` by 5-6 folds. Our estimate for the disk space consumption for the entire experiment (full-scale) is 800-850 MB currently. Without this step the estimate would shoot up to around 6-7 GB for a full-scale experiment.

Reply:
> > According to my readings, this step reduces the disk space consumption by about 5-6 folds.
> 
> Right, but how big is a file without this compression vs with it?
> If we are going from 80 MB in size for a full experiment to 20 MB, I don't think it is worth it.
> 
> I'm also concerned with this method you've used. It isn't sound. There can be collisions here if two things happen to have the same 8 char prefix. Why can't we simply use a numerical ID?

Since we call this method in each thread (multiprocessing), we use hashing to resolve names to ids as we want to have consistent IDs for the same names in different threads and hashing is a good way to do it.

Moreover, we call `name_to_id` and `remove_redundant_duplicates` in each thread as this helps us reduce the memory consumption (as it reduces the df size frequently). We can use numerical ID but then we would have to resolve names to IDs after we collect all the data for the experiment which would increase the memory consumption considerably. While hashing there is possibility of collision of IDs, no doubt, but the probability is quite low. So, currently I'm truncating the computed hash at 7 characters but we can increase this to say 10 or 12 characters which would drop the probability of collision further.

Also for my local experiments (not full-scale), using `name_to_id` helps reduce disk space consumed by the *.csv.gz by 5-6 folds. Our estimate for the disk space consumption for the entire experiment (full-scale) is 800-850 MB currently. Without this step the estimate would shoot up to around 6-7 GB for a full-scale experiment.